### PR TITLE
Implement initial API

### PR DESCRIPTION
**Pull request by marcoskwkm**
#### What problem is this solving?

This implements `OrderItemsProvider`, `useOrderItems`, as well as the API methods `itemList` and `updateItem`.

#### How should this be manually tested?

Go to [this workspace](https://orderitems--vtexgame1.myvtex.com), add some items to the cart then visit `/cart`. A modified `product-list` which makes use `OrderItems` is linked to this workspace.

#### Checklist/Reminders

- [x] Updated `README.md`.
- [x] Updated `CHANGELOG.md`.
- [x] [Linked this PR to a Clubhouse story (if applicable)](https://app.clubhouse.io/vtex/story/18344/implementar-lista-de-items-e-updateitems).
- [x] Updated/created tests (important for bug fixes).
- [ ] Deleted the workspace after merging this PR (if applicable).

#### Type of changes

✔️ | Type of Change
---|---
_ | Bug fix 
✔️ | New feature 
_ | Breaking change 
_ | Technical improvements 

#### Notes

The component maintains the list of items in a state in order to implement an optimistic UI. This is not ideal because it causes inconsistencies between multiple instances of this component. This was made just to facilitate testing the implementation. The idea is to later move this state to `OrderManager`.

### CODE REVIEW

File: react/OrderItems.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  UpdateItem: any
+}
+
+const OrderItemsContext = createContext<Context | undefined>(undefined)
+
+const EmptyState: FunctionComponent = ({ children }: any) => {
+  const updateItem = async (_: number, __: number) => {}
+  return (
+    <OrderItemsContext.Provider
+      value={{ itemList: [], updateItem, loading: true }}
```
Comment (klzns) [inline react/OrderItems.tsx]: When filling context Provider value and its value it's an object, use useMemo to memoize the object.

```tsx
const value = useMemo(() => ({ itemList: [], updateItem, loading: true }), [])

return (
  <OrderItemsContext.Provider value={value}>
```

File: react/OrderItems.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    >
+      {children}
+    </OrderItemsContext.Provider>
+  )
+}
+
+let lastUpdateToken = { outdated: false }
```
Comment (klzns) [inline react/OrderItems.tsx]: Why is that here? Why not use useRef? Keeping it here will lead to bugs if there are two instances of this provider

File: react/OrderItems.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    enqueue(task).catch((error: any) => {
+      if (!error || error.code !== 'TASK_CANCELLED') {
+        throw error
+      }
+    })
+  }
+
+  return (
+    <OrderItemsContext.Provider
+      value={{ itemList, updateItem: updateItem, loading: false }}
```
Comment (klzns) [inline react/OrderItems.tsx]: Same point as above

File: react/__tests__/OrderItems.test.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    const { getByText } = render(<OuterComponent />, {
+      graphql: { mocks: [ItemListMock, UpdateItemMock] },
+    })
+    await new Promise(resolve => setTimeout(() => resolve())) // waits for initial item query
+
+    const button = getByText('mutate')
+    fireEvent.click(button)
+    expect(getByText(`${mockItems[1].name}: 123`)).toBeTruthy() // optimistic response
+
+    await new Promise(resolve => setTimeout(() => resolve())) // waits for actual mutation result
```
Comment (klzns) [inline react/__tests__/OrderItems.test.tsx]: We created the flushPromises API in test-tools, check it out

File: react/OrderItems.tsx (hunk 2 above)
Comment (marcoskwkm) [reply react/OrderItems.tsx]: Changed to a much better approach which is to use `OrderManager`'s queue status instead as you suggested.

File: react/__tests__/OrderItems.test.tsx (hunk 4 above)
Comment (marcoskwkm) [reply react/__tests__/OrderItems.test.tsx]: 😮

File: react/OrderItems.tsx [hunk 5]
```diff
@@ -48,53 +48,70 @@ export const OrderItemsProvider = compose(
...
+  useEffect(() => {
+    const unlisten = listen('Pending', () => (isQueueBusy.current = true))
+    return unlisten
+  })
+  useEffect(() => {
+    const unlisten = listen('Fulfilled', () => (isQueueBusy.current = false))
+    return unlisten
+  })
+
+  const updateItem = useMemo(
```
Comment (klzns) [inline react/OrderItems.tsx]: You can use useCallback instead

It's just like useMemo, but to create memoized functions

File: react/OrderItems.tsx [hunk 6]
```diff
@@ -48,53 +48,70 @@ export const OrderItemsProvider = compose(
...
+    () => (index: number, quantity: number) => {
+      const updatedList =
+        quantity === 0
+          ? [...itemList.slice(0, index), ...itemList.slice(index + 1)]
+          : adjust(index, item => ({ ...item, quantity }), itemList)
+
+      setItemList(updatedList)
```
Comment (klzns) [inline react/OrderItems.tsx]: That's for optmistic update, right? 

I think we should use the GraphQL API for that instead:
https://www.apollographql.com/docs/react/features/optimistic-ui/

That way, the query will respond immediately

Comment (marcoskwkm) [reply react/OrderItems.tsx]: I considered this, but doesn't work because the request will be enqueued, so the "optimistic response" will only show up when this task starts running.

File: react/OrderItems.tsx (hunk 5 above)
Comment (marcoskwkm) [reply react/OrderItems.tsx]: TIL how to react hooks

File: react/package.json [hunk 7]
```diff
@@ -22,7 +22,7 @@
     "@types/react": "^16.8.10",
     "@types/react-intl": "^2.3.17",
     "@types/recompose": "^0.30.7",
-    "@vtex/test-tools": "^1.1.0",
+    "@vtex/test-tools": "^0.2.0",
```
Comment (klzns) [inline react/package.json]: Ops, keep that

Comment (marcoskwkm) [reply react/package.json]: Fine =P
